docs/designs/allreduce_design.md

```diff
@@ -0,0 +1,64 @@
+# Design for Allreduce Support
+
+## Existing Collective Communication Technologies
+
+The following three libraries provide collective communications and all of them have been adopted by large projects:
+
+* [Gloo](https://github.com/facebookincubator/gloo/)
+* [NCCL](https://github.com/NVIDIA/nccl)
+* [Rabit](https://github.com/dmlc/rabit)
+
+### Gloo
+
+Gloo is a collective communications library. It comes with a number of collective algorithms useful for machine learning
+applications, which includes but not limited to Broadcast and Allreduce. It has been adopted by [PyTorch](https://github.com/pytorch/pytorch).
+
+Transport of data between participating machines is abstracted so that IP can be used at all times, or InifiniBand (or RoCE)
+when available. In the latter case, if the InfiniBand transport is used, GPUDirect can be used to accelerate cross machine
+GPU-to-GPU memory transfers. Gloo includes several collective algorithm implementations that work directly with NVIDIA GPU buffers.
+These take advantage of overlapping host and GPU operations to decrease overall latency.
+
+Gloo does not support fault tolerance but supports both GPUs and CPUs for at least Allreduce and Broadcast primitives.
+
+### NCCL
+
+NCCL (NVIDIA Collective Communications Library) is a library of multi-GPU collective communication primitives that are topology-aware and
+can be easily integrated into applications. It has been adopted by both [TensorFlow](https://github.com/tensorflow/tensorflow/) and [PyTorch](https://github.com/pytorch/pytorch).
+
+NCCL focuses on accelerating collective communication primitives. For example, NCCL conveniently removes
+the need for developers to optimize their applications for specific machines. In addition, NCCL provides fast collectives
+over multiple GPUs both within and across nodes. It supports a variety of interconnect technologies including PCIe, NVLINK,
+InfiniBand Verbs, and IP sockets. NCCL also automatically patterns its communication strategy to match the system’s underlying
+GPU interconnect topology.
+
+NCCL does not support fault tolerance but one can support this by filtering out the failed workers, reassigning ranks, and then
+reconstructuring the [NCCL Communicator](https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/usage/communicators.html).
+Also note that NCCL only supports GPUs for the collective communication primitives.
```

Comment:
A reference blog which tests the elastic scheduling of NCCL: https://my.oschina.net/u/1459307/blog/1650028
An interesting observation is that: if one process of the NCCL group is killed, the other process in the group will be blocked, both CPU and GPU usage are 100%. It seems that we could only kill all the processes in NCCL group in force to stop the job.

Reply:
Thanks. The blog post seems old and is in Chinese so I am not including it in the doc this time.